deliver/lib/deliver/upload_metadata.rb

```diff
@@ -427,25 +427,27 @@ def detect_languages(options)
         .uniq
     end
 
-    def fetch_edit_app_store_version(app, platform, wait_time: 10)
-      retry_if_nil("Cannot find edit app store version", wait_time: wait_time) do
+    def fetch_edit_app_store_version(app, platform, max_retries:, initial_wait_time: 10)
+      retry_if_nil("Cannot find edit app store version", tries: max_retries, initial_wait_time: initial_wait_time) do
         app.get_edit_app_store_version(platform: platform)
       end
     end
 
-    def fetch_edit_app_info(app, wait_time: 10)
-      retry_if_nil("Cannot find edit app info", wait_time: wait_time) do
+    def fetch_edit_app_info(app, max_retries:, initial_wait_time: 10)
+      retry_if_nil("Cannot find edit app info", tries: max_retries, initial_wait_time: initial_wait_time) do
         app.fetch_edit_app_info
       end
     end
 
-    def fetch_live_app_info(app, wait_time: 10)
-      retry_if_nil("Cannot find live app info", wait_time: wait_time) do
+    def fetch_live_app_info(app, max_retries:, initial_wait_time: 10)
+      retry_if_nil("Cannot find live app info", tries: max_retries, initial_wait_time: initial_wait_time) do
         app.fetch_live_app_info
       end
     end
 
-    def retry_if_nil(message, tries: 5, wait_time: 10)
+    # Retries a block of code if the return value is nil, with an exponential backoff.
+    def retry_if_nil(message, tries:, initial_wait_time: 10)
```

Comment:
CI will timeout in this retry method because of `no console output` if tries are more because we are doubling seconds after every retry.
A better solution could be that keep printing the retry status on the console to avoid CI timeout. Please see the attached modified version with the output. cc: @rogerluan 

```ruby
  def retry_if_nil(message, tries:, initial_wait_time: 10)
    wait_time = initial_wait_time
    loop do
      tries -= 1

      value = yield
      return value if value
      UI.message("#{message}... Retrying after #{wait_time} seconds (remaining: #{tries} tries)")
      
      retrying_status_time = 0
      while retrying_status_time < wait_time  do
        UI.message("Retrying status: #{retrying_status_time}/#{wait_time} seconds")
        sleep(initial_wait_time)
        retrying_status_time += initial_wait_time
      end

      return nil if tries.zero?

      wait_time *= 2 # Double the wait time for the next iteration
    end
  end
  ```
  
 #### Output:
<img width="968" alt="Screenshot 2024-01-06 at 00 19 48" src="https://github.com/fastlane/fastlane/assets/5364500/2f0fac09-c4b1-40ef-992f-006cbb511eb9">

Reply:
This is a very good point @crazymanish ! 

I'm torn between your implementation (which's great IMO), or incorporating the feedback discussed earlier in this PR about trying exponentially only until a certain threshold (e.g. 5 minutes), and then from then on keep trying every 5 minutes. As long as this threshold is <10minutes (which is the default console log timeout used in many CI services), we should be good.

What do you think? Should we incorporate your code regardless of the decision above? There's no harm in it, just makes code more complicated (justifiably so)

Reply:
Aha! i missed the threshold comment...Any retrying logic before the CI timeout (10 minutes) sounds good...
we should be good as long as CI will not timeout and CI will respect `DELIVER_VERSION_CHECK_WAIT_RETRY_LIMIT` even if the retry_limit value is long i.e 10 or 15 retries

-----

Another small thing we could do is: we can **have the customized threshold value** This will give freedom to our users to retry after every 1 minute, 3 minutes, 5 minutes, or any X minutes (based on their CI timeout X values)
- Define `DELIVER_VERSION_CHECK_WAIT_RETRY_THRESHOLD` or `version_check_wait_retry_threshold` param
- Make this threshold by default every 3 minutes to avoid CI timeout and it should be easy to implement too
```ruby 
wait_time = min(wait_time * 2, version_check_wait_retry_threshold * 60)`


btw, Thank you very much for doing the good work in this PR, it seems lots of people are facing this issue!!

Reply:
Thanks for the suggestion @crazymanish 🙏 I usually try to avoid bloating up the actions/tools with so many options, and I think in this case there's no need to allow _that_ level of customization. I'd go with that implementation and go for a safe fixed number of 5 minutes which seems to be reasonable and work with every CI provider I can think of 👀 do you feel strongly we should allow customization of this by the users?

I'm going to go ahead and implement this now but I can change later too 🙏